openff/toolkit/topology/mm_molecule.py

```diff
@@ -0,0 +1,475 @@
+            return 0
+        return len(self.conformers)
+
+    def atom(self, index):
+        return self.atoms[index]
+
+    def atom_index(self, atom) -> int:
+        return self.atoms.index(atom)
+
+    def bond(self, index):
+        return self.bonds[index]
+
+    def get_bond_between(self, atom1_index, atom2_index):
+        atom1 = self.atom(atom1_index)
+        atom2 = self.atom(atom2_index)
+        for bond in atom1.bonds:
+            for atom in bond.atoms:
+                if atom is atom2:
+                    return bond
+
+    def particle(self, index) -> int:
+        return self.atom(index)
+
+    @property
+    def particles(self):
+        return self.atoms
+
+    def particle_index(self, particle) -> int:
+        return self.atom_index(particle)
+
+    @property
+    def angles(self):
+        for atom1 in self.atoms:
+            for atom2 in atom1.bonded_atoms:
+                for atom3 in atom2.bonded_atoms:
+                    if atom1 is atom3:
+                        continue
+                    if atom1.molecule_atom_index < atom3.molecule_atom_index:
+                        yield (atom1, atom2, atom3)
+                    else:
+                        # Do not return i.e. (2, 1, 0), only (0, 1, 2)
+                        pass
+
+    @property
+    def propers(self):
+        for atom1 in self.atoms:
+            for atom2 in atom1.bonded_atoms:
+                for atom3 in atom2.bonded_atoms:
+                    if atom1 is atom3:
+                        continue
+                    for atom4 in atom3.bonded_atoms:
+                        if atom4 in (atom1, atom2):
+                            continue
+
+                        if atom1.molecule_atom_index < atom4.molecule_atom_index:
+                            yield (atom1, atom2, atom3, atom4)
+                        else:
+                            # Do no duplicate
+                            pass  # yield (atom4, atom3, atom2, atom1)
+
+    @property
+    def impropers(self):
+        return {}
+
+    @property
+    def smirnoff_impropers(self):
+        return {}
+
+    @property
+    def amber_impropers(self):
+        return {}
+
+    @property
+    def hill_formula(self) -> str:
+        """
+        Return the Hill formula of this molecule.
+
+        If any atoms are non-elemental, defined by having an atomic number of 0,
+        return INVALID instead.
+        """
+        return self.to_hill_formula()
+
+    def to_hill_formula(self) -> str:
+        atom_nums: List[int] = [atom.atomic_number for atom in self.atoms]
+
+        if min(atom_nums) <= 0:
+            return "INVALID"
+
+        return _atom_nums_to_hill_formula(atom_nums)
+
+    def to_networkx(self) -> "nx.Graph":
+        import networkx as nx
+
+        graph = nx.Graph()
+
+        for atom_index, atom in enumerate(self.atoms):
+            graph.add_node(
+                atom_index,
+                atomic_number=atom.atomic_number,
```

Comment:
We should attach any custom attributes that users have added to atoms or bonds to the nodes and edges. 

If possible, we should do the same in `from_networkx`

Reply:
https://github.com/openforcefield/openff-toolkit/pull/1179/commits/a8329aa2515cc00a1efb869f03480e30fe078fcc